[chips] Block all third-party cookies when CHIPS is disabled

sysrqb opened the pull request:
#### f7cd5bc469411f0807f10d7212fb997c7f9052ca
<pre>
[chips] Block all third-party cookies when CHIPS is disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=296892">https://bugs.webkit.org/show_bug.cgi?id=296892</a>
<a href="https://rdar.apple.com/157495921">rdar://157495921</a>

Reviewed by Sihui Liu.

Currently, we don&apos;t fully enforce third-party cookie blocking when the CHIPS
compile-time flag is true but the run-time flag is false. The expected behavior
is that we would fallback to using the shipping ITP behavior, where we block
all third-party cookies, without regard for partitioned cookies.

This patch adjusts the logic so that we block all third party cookies unless
third-party cookie blocking is disabled. It also simplifies the code by
combining duplicate code into NetworkStorageSession::shouldBlockCookies.

This is covered by a currently skipped tests, and some tests that will be
included in a future PR. We can&apos;t currently test this behavior because CHIPS is
disabled at compile-time.

* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::thirdPartyCookieBlockingDecisionForRequest const):
(WebCore::NetworkStorageSession::shouldBlockCookies const):
(WebCore::NetworkStorageSession::shouldBlockCookies):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::requestThirdPartyCookieBlockingDecision const):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkTaskCocoa::shouldBlockCookies): Deleted.

Canonical link: <a href="https://commits.webkit.org/298695@main">https://commits.webkit.org/298695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b40109a539074f8a6726038ce413b46651ecc35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116340 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26550 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122397 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66901 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118229 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44589 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88368 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119289 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104380 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68799 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28350 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22483 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66078 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98650 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22644 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125546 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32468 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97075 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43599 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100586 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96871 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42159 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20055 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39186 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18583 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43121 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48713 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42588 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45928 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44293 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -216,14 +216,23 @@ ThirdPartyCookieBlockingDecision NetworkStorageSession::thirdPartyCookieBlocking
+bool NetworkStorageSession::shouldBlockCookies(ThirdPartyCookieBlockingDecision thirdPartyCookieBlockingDecision) const
+{
+#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
+    if (!isOptInCookiePartitioningEnabled())
```
I remember when `isOptInCookiePartitioningEnabled()` is false,  `thirdPartyCookieBlockingDecision` can only be `All` or `None`, so isn't `thirdPartyCookieBlockingDecision != WebCore::ThirdPartyCookieBlockingDecision::None` the same as `thirdPartyCookieBlockingDecision == ThirdPartyCookieBlockingDecision::All`?

sysrqb replied:
Not exactly. When `isOptInCookiePartitioningEnabled()` is false, we can be in the mode  `ThirdPartyCookieBlockingMode::AllExceptPartitioned`. That state isn't available when the compile-time guard is false, but when the compile-time guard is true then `AllExceptPartitioned` is a valid mode. This is necessary so we can switch between `isOptInCookiePartitioningEnabled()` being true or false and have the correct behavior. We could change the behavior so that `AllExceptPartitioned` is not valid when `isOptInCookiePartitioningEnabled()` is false, but then we would need some additional logic to cache the previously enabled state and enforce this invariant.

szewai replied:
> but when the compile-time guard is true then AllExceptPartitioned is a valid mode.

Even the mode option exists, I don't think we would get that option if `isOptInCookiePartitioningEnabled` is false? E.g. `NetworkStorageSession::thirdPartyCookieBlockingDecisionForRequest` only returns `ThirdPartyCookieBlockingDecision::AllExceptPartitioned` if the flag is true.
So it seems if the flag is false and mode is `AllExceptPartitioned` then there is something wrong in the implementation. (If we cache the blocking mode somewhere, then the mode should be updated when the runtime flag is updated.)

sysrqb replied:
We don't currently check the flag here:
```
    switch (m_thirdPartyCookieBlockingMode) {
    case ThirdPartyCookieBlockingMode::All:
        return ThirdPartyCookieBlockingDecision::All;
 [...]
#if ENABLE(OPT_IN_PARTITIONED_COOKIES)
    case ThirdPartyCookieBlockingMode::AllExceptPartitioned:
        return ThirdPartyCookieBlockingDecision::AllExceptPartitioned;
#endif
```

You are recommending that we should check the flag in this last case and return `::All`? That's an alternative solution. You prefer that?

szewai replied:
Yes, that seems consistent with the existing implementation.

Commit:
[chips] Block all third-party cookies when CHIPS is disabled
https://bugs.webkit.org/show_bug.cgi?id=296892
rdar://157495921

Reviewed by Sihui Liu.

Currently, we don't fully enforce third-party cookie blocking when the CHIPS
compile-time flag is true but the run-time flag is false. The expected behavior
is that we would fallback to using the shipping ITP behavior, where we block
all third-party cookies, without regard for partitioned cookies.

This patch adjusts the logic so that we block all third party cookies unless
third-party cookie blocking is disabled. It also simplifies the code by
combining duplicate code into NetworkStorageSession::shouldBlockCookies.

This is covered by a currently skipped tests, and some tests that will be
included in a future PR. We can't currently test this behavior because CHIPS is
disabled at compile-time.

* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::thirdPartyCookieBlockingDecisionForRequest const):
(WebCore::NetworkStorageSession::shouldBlockCookies const):
(WebCore::NetworkStorageSession::shouldBlockCookies):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::requestThirdPartyCookieBlockingDecision const):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkTaskCocoa::shouldBlockCookies): Deleted.

Canonical link: https://commits.webkit.org/298695@main
